### PR TITLE
feat: wrap root_dir in coroutine.

### PR DESCRIPTION
**Pull request by glepnir**
1. make get `root_dir` can be async wrap in coroutine
2. change rust_analyzer get command output async.

### CODE REVIEW

File: lua/lspconfig/configs.lua [hunk 1]
```diff
@@ -257,16 +255,25 @@ function configs.__newindex(t, config_name, config_def)
       end
       local buf_path = util.path.sanitize(bufname)
 
-      if get_root_dir then
-        root_dir = get_root_dir(buf_path, bufnr)
-      end
+      local co = coroutine.create(function()
```
Comment (ii14) [inline lua/lspconfig/configs.lua:258]: `coroutine.resume(coroutine.create(...))`? `co` won't be used anywhere else

**Review by ii14 — COMMENTED** [review]
`root_dir` is also called in `M.launch`, that has to be wrapped in a coroutine too.

Not sure what to do about `:LspInfo`, `root_dir` function is also called there

**Comment by glepnir** [comment]
`M.launch` is also invoked `manager.add` . so it's fine. in test `LspRestart` and`LspStart` works fine.

File: lua/lspconfig/server_configurations/rust_analyzer.lua [hunk 2]
```diff
@@ -10,47 +11,77 @@ local function reload_workspace(bufnr)
...
+    if not uv.is_closing(handle) then
+      uv.close(handle)
+    end
+  end
+
+  local stdout = uv.new_pipe(false)
+  local stderr = uv.new_pipe(false)
+  local chunks = {}
+  local handle
+  handle = uv.spawn('cargo', {
```
[the remark this answers is not in the document]
Comment (glepnir) [reply lua/lspconfig/server_configurations/rust_analyzer.lua:30]: jobstart is like a syntactic sugar based on uv.spawn. But something is done internally to make it easy to use. The current use case is simple though. So using uv.spawn directly can simplify the execution of some wrapping syntax. Although the time is negligible.

**Comment by ii14** [comment]
`launch` function is used directly by `:LspStart` and `:LspRestart`. `get_workspace_dir` simply won't work when using those commands.

**Comment by glepnir** [comment]
hmmm I test in a helloword rust project. `LspStop` then `LspStart` `get_workspace_dir` is run twice i get two print.  add print before` coroutin.resume(co, workspace_root)` line:46

File: lua/lspconfig/server_configurations/rust_analyzer.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (ii14) [reply lua/lspconfig/server_configurations/rust_analyzer.lua:30]: `uv.spawn` is more verbose and harder to use correctly, and `jobstart` fixes some of its pitfalls. This code will also be an example of how to use it, so I'd just use the simpler alternative.

Also, failure has to be handled here. Looks like it falls back to other ways of determining the root directory. And if starting the process fails, the coroutine will be suspended and never resumed, not giving other code a chance to run.

**Comment by ii14** [comment]
It will abort the function here:
```lua
  local co = coroutine.running()
  if not co then
    return
  end
```
`coroutine.running()` returns nil when its called outside of coroutines

**Comment by glepnir** [comment]
get root coroutine in `manager.add` this function in `M.launch`

**Comment by ii14** [comment]
Okay, but `root_dir` function is also called in configs.lua on line 110. That part won't be correct. We want to give a guarantee that `root_dir` functions are *always* called inside a coroutine. Try adding `assert(coroutine.running())` at the beginning of the `get_workspace_dir` function.

**Comment by ii14** [comment]
Actually I guess such guarantee wouldn't be really necessary, but in that case you need to handle both sync and async context inside `root_dir` functions.

**Comment by glepnir** [comment]
okay . i need think about that carefully.

**Comment by ii14** [comment]
Don't forget about `:LspInfo`, it calls `root_dir` too, and everything is synchronous there. The way I see it the solutions are:

1. Make the entire `:LspInfo` asynchronous. You call `:LspInfo`, window shows up once all results are collected.
2. Show the window instantly with anything that's already available, update the buffer as you get back results asynchronously.
3. Don't guarantee that `root_dir` is always a coroutine, handle both sync/async it `root_dir` functions and leave `:LspInfo` as it is - blocking.

**Comment by glepnir** [comment]
root dir in `LspInfo`  we can get it from `client.config.root_dir` this is more simply we don't need call root_dir function.

**Comment by ii14** [comment]
That's what I initially thought too, but it's for the `Other clients that match the filetype` section, where it says what clients can be started, but are not necessarily actually started yet.

**Comment by glepnir** [comment]
range all server name keys in configs. get the all the client that filetypes match current filetype. if client can get from active clients then root dir is `client.config.root_dir` . if not this server not start. then root dir is `not found`.

**Comment by glepnir** [comment]
maybe need abstract this `coroutine.resume(coroutine.create)` like `if get_root_dir func then ` invoke a wrap function and pass get_root_dir into with a callback.   lspinfo and launch had changed. how do you think ?

File: lua/lspconfig/ui/lspinfo.lua [hunk 3]
```diff
@@ -64,9 +64,12 @@ local function make_config_info(config, bufnr)
...
   end)
-  local root_dir = config.get_root_dir and config.get_root_dir(buffer_dir)
-  if root_dir then
-    config_info.root_dir = root_dir
+
+  if config.get_root_dir then
+    config_info.root_dir = coroutine.resume(coroutine.create(vim.schedul_wrap(function()
+      local root_dir = config.get_root_dir(buffer_dir)
+      coroutine.yield(root_dir)
+    end)))
```
Comment (ii14) [inline lua/lspconfig/ui/lspinfo.lua:72]: If `get_root_dir` calls `coroutine.yield` first, the result of that will be returned from this `coroutine.resume` (and as the second value, the first value is the status of the coroutine) and assigned to `root_dir`, which wouldn't be correct. We specifically want the result of `config.get_root_dir`.

Also there is a typo in `schedul_wrap`

File: lua/lspconfig/configs.lua [hunk 4]
```diff
@@ -97,54 +97,54 @@ function configs.__newindex(t, config_name, config_def)
...
-          if not util.bufname_valid(bufname) then
-            return
-          end
+      local bufnr = api.nvim_get_current_buf()
+      local bufname = api.nvim_buf_get_name(bufnr)
+      if (#bufname == 0 and not config.single_file_support) or (#bufname ~= 0 and not util.bufname_valid(bufname)) then
+        return
+      end
+
+      coroutine.resume(coroutine.create(vim.schedule_wrap(function()
```
[the remark this answers is not in the document]
Comment (glepnir) [reply lua/lspconfig/configs.lua:106]: because if don't run it on main thread it just start a server but not attach buffer.  maybe something wrong on lsp.buf_attach_client. I need check.

File: lua/lspconfig/ui/lspinfo.lua (hunk 3 above)
Comment (glepnir) [reply lua/lspconfig/ui/lspinfo.lua:72]: yep first value is status. a mistake. but i really don't like this in lspinfo ..

Comment (glepnir) [reply lua/lspconfig/ui/lspinfo.lua]: how about now

File: lua/lspconfig/ui/lspinfo.lua [hunk 5]
```diff
@@ -64,9 +64,13 @@ local function make_config_info(config, bufnr)
...
-  local root_dir = config.get_root_dir and config.get_root_dir(buffer_dir)
-  if root_dir then
-    config_info.root_dir = root_dir
+
+  if config.get_root_dir then
+    local status, root_dir = coroutine.resume(coroutine.create(vim.schedule_wrap(function()
+      local root_dir = config.get_root_dir(buffer_dir)
+      coroutine.yield(root_dir)
+    end)))
+    config_info.root_dir = status and root_dir or error_messages.root_dir_not_found
```
Comment (ii14) [inline lua/lspconfig/ui/lspinfo.lua:73]: Let's not use `coroutine.yield` outside of the "user code" and pass the result through a reference/upvalue. If for some reason the initial suspend is something like `coroutine.yield(42)`, `root_dir` will be assigned to 42. Not that anyone should do that, but a reference will be more robust.
```suggestion
    local root_dir
    local status = coroutine.resume(coroutine.create(vim.schedule_wrap(function()
      root_dir = config.get_root_dir(buffer_dir)
    end)))
    config_info.root_dir = status and root_dir or error_messages.root_dir_not_found
```
But doing this alone will make it stop working for anything has an async `root_dir` function. Preferably the buffer could be updated with this information once we actually get the result. Maybe checking if `coroutine.status` is `'suspended'`, and if so then save the line number and update that line? But this will make the code more complicated.

Comment (glepnir) [reply lua/lspconfig/ui/lspinfo.lua]: > Maybe checking if coroutine.status is 'suspended', and if so then save the line number and update that line? But this will make the code more complicated.

honestly yes .. I don't have good idea on lspinfo .  It has some importance. But not that important.

File: lua/lspconfig/configs.lua [hunk 6]
```diff
@@ -257,16 +257,23 @@ function configs.__newindex(t, config_name, config_def)
...
-      if root_dir then
-        manager.add(root_dir, false, bufnr)
-      elseif config.single_file_support then
-        local pseudo_root = #bufname == 0 and uv.cwd() or util.path.dirname(buf_path)
-        manager.add(pseudo_root, true, bufnr)
-      end
+        if root_dir then
+          vim.schedule(function()
+            manager.add(root_dir, false, bufnr)
+          end)
```
Comment (ii14) [inline lua/lspconfig/configs.lua:269]: Avoid `vim.schedule` when it's not necessary? Same on the branch below
```suggestion
          if not vim.in_fast_event() then
            manager.add(root_dir, false, bufnr)
          else
            vim.schedule(function()
              manager.add(root_dir, false, bufnr)
            end)
          end
```
Or it could be even something like this, maybe even a new utility function that does this:
```lua
local function cb()
  manager.add(root_dir, false, bufnr)
end
if vim.in_fast_event() then
  vim.schedule(cb)
else
  cb()
end
```

File: lua/lspconfig/server_configurations/rust_analyzer.lua [hunk 7]
```diff
@@ -10,6 +10,44 @@ local function reload_workspace(bufnr)
   end)
 end
 
+local function get_workspace_dir(cmd)
+  assert(coroutine.running())
+  local co = coroutine.running()
+  if not co then
+    return
+  end
```
Comment (ii14) [inline lua/lspconfig/server_configurations/rust_analyzer.lua:18]:
```suggestion
  local co = assert(coroutine.running())
```

File: lua/lspconfig/configs.lua (hunk 6 above)
Comment (glepnir) [reply lua/lspconfig/configs.lua:269]: make sense

File: lua/lspconfig/configs.lua [hunk 8]
```diff
@@ -257,16 +257,29 @@ function configs.__newindex(t, config_name, config_def)
       end
       local buf_path = util.path.sanitize(bufname)
 
-      if get_root_dir then
-        root_dir = get_root_dir(buf_path, bufnr)
+      local function check_fast(root_dir)
+        if not vim.in_fast_event() then
+          manager.add(root_dir, false, bufnr)
```
Comment (ii14) [inline lua/lspconfig/configs.lua:262]: Second argument can't be false all the time, it used to be true on the `config.single_file_support` branch

File: lua/lspconfig/ui/lspinfo.lua [hunk 9]
```diff
@@ -64,9 +64,13 @@ local function make_config_info(config, bufnr)
   local buffer_dir = api.nvim_buf_call(bufnr, function()
     return vim.fn.expand '%:p:h'
   end)
-  local root_dir = config.get_root_dir and config.get_root_dir(buffer_dir)
-  if root_dir then
-    config_info.root_dir = root_dir
+
+  if config.get_root_dir then
+    local root_dir
+    local status = coroutine.resume(coroutine.create(vim.schedule_wrap(function()
```
Comment (ii14) [inline lua/lspconfig/ui/lspinfo.lua:70]: I still don't get why `vim.schedule_wrap` is here, I think it's unnecessary and it only forces `get_root_dir` to be deferred. If I'm not mistaken this makes it so `root_dir` is always going to be `nil`, because it's assigned on the next event loop iteration once all of this code is done

Comment (glepnir) [reply lua/lspconfig/ui/lspinfo.lua:70]: > I think it's unnecessary and it only forces get_root_dir to be deferred

that's right ..looks like before i don't get it work then i add it into schedule wrap . let me try again

**Comment by ii14** [comment]
What's really important with coroutines is error handling. `coroutine.resume` works kinda like `pcall`, it returns the status and error message. If we don't handle errors they are just going to be silent and if something goes wrong we won't have any way of debugging it.

To make sure we are catching all unhandled errors (except in callbacks before they resume the coroutine, but that should be handled by nvim already), the coroutine body should be executed inside `pcall` like this:
```lua
local function func()
  error('test')
end

coroutine.resume(coroutine.create(function()
  local status, err = pcall(function()
    func()
  end)
  if not status then
    print('lspconfig: unhandled error: ' .. tostring(err))
  end
end))
```

**Comment by glepnir** [comment]
yep use `print` can works well in coroutine thread. if use notify need yield error . or use a wrap try catch

```lua
function tryCatch(tryFunc, catchFunc)
    local success, result = pcall(tryFunc)
    if not success then
        catchFunc(result)
    end
end

tryCatch(function()
    error("xxxx")
end, function(err)
    print("Lspconfig Error: ", err)
end)
```

File: lua/lspconfig/configs.lua [hunk 10]
```diff
@@ -72,17 +72,10 @@ function configs.__newindex(t, config_name, config_def)
...
-      if config.filetypes then
-        event = 'FileType'
-        pattern = table.concat(config.filetypes, ',')
-      else
-        event = 'BufReadPost'
-        pattern = '*'
-      end
-      api.nvim_create_autocmd(event, {
-        pattern = pattern,
+      local event_conf = config.filetypes and { event = 'FileType', pattern = table.concat(config.filetypes, ',') }
```
Comment (ii14) [inline lua/lspconfig/configs.lua:75]: `pattern` can be a table
```suggestion
      local event_conf = config.filetypes and { event = 'FileType', pattern = config.filetypes }
```

File: lua/lspconfig/configs.lua [hunk 11]
```diff
@@ -97,54 +90,57 @@ function configs.__newindex(t, config_name, config_def)
...
-        -- We just no longer send rootDirectory or workspaceFolders during initialization.
-        local bufname = api.nvim_buf_get_name(0)
-        if #bufname ~= 0 and not util.bufname_valid(bufname) then
-          return
-        end
-        local pseudo_root = #bufname == 0 and uv.cwd() or util.path.dirname(util.path.sanitize(bufname))
-        M.manager.add(pseudo_root, true, api.nvim_get_current_buf())
       end
+
+      util.async_run(try_launch)
```
Comment (ii14) [inline lua/lspconfig/configs.lua:143]: Can't `try_launch` be inlined here?
```lua
util.async_run(function()
  -- ...
end)
```

File: lua/lspconfig/configs.lua [hunk 12]
```diff
@@ -255,35 +250,48 @@ function configs.__newindex(t, config_name, config_def)
...
+        end
+
+        if root_dir then
+          check_fast(root_dir, false)
+        elseif config.single_file_support then
+          local pseudo_root = #bufname == 0 and pwd or util.path.dirname(buf_path)
+          check_fast(pseudo_root, true)
+        end
       end
+      util.async_run(try_start)
```
Comment (ii14) [inline lua/lspconfig/configs.lua:284]: Same here, `try_start` can be inlined

File: lua/lspconfig/server_configurations/rust_analyzer.lua [hunk 13]
```diff
@@ -10,6 +10,40 @@ local function reload_workspace(bufnr)
...
+      local data = table.concat(chunks, '')
+      chunks = vim.json.decode(data)
+      local workspace_root = chunks and chunks['workspace_root'] or nil
+      coroutine.resume(co, workspace_root)
+    end,
+    stdout_buffered = true,
+    stderr_buffered = true,
+  })
+
+  if jobid < 0 then
```
Comment (ii14) [inline lua/lspconfig/server_configurations/rust_analyzer.lua:39]: Zero is failure too
```suggestion
  if jobid <= 0 then
```

File: lua/lspconfig/server_configurations/rust_analyzer.lua [hunk 14]
```diff
@@ -10,6 +10,40 @@ local function reload_workspace(bufnr)
   end)
 end
 
+local function get_workspace_dir(cmd)
+  local co = assert(coroutine.running())
+
+  local chunks = {}
+  local jobid = vim.fn.jobstart(cmd, {
+    on_stdout = function(_, data, _)
+      chunks[#chunks + 1] = unpack(data)
```
Comment (ii14) [inline lua/lspconfig/server_configurations/rust_analyzer.lua:19]: `var = unpack(data)` is equivalent to `var = data[1]`, other values are discarded. `data` array is split on newlines, so I think that should work:
```suggestion
      chunks[#chunks + 1] = table.concat(data, '\n')
```

File: lua/lspconfig/server_configurations/rust_analyzer.lua [hunk 15]
```diff
@@ -10,6 +10,40 @@ local function reload_workspace(bufnr)
...
+    end,
+    on_exit = function()
+      if next(chunks) == nil then
+        coroutine.resume(co, nil)
+        return
+      end
+
+      local data = table.concat(chunks, '')
+      chunks = vim.json.decode(data)
+      local workspace_root = chunks and chunks['workspace_root'] or nil
```
Comment (ii14) [inline lua/lspconfig/server_configurations/rust_analyzer.lua:32]: Can you move all of this out of the callback function, after `coroutine.yield()` and just call `coroutine.resume(co)` here?

File: lua/lspconfig/ui/lspinfo.lua [hunk 16]
```diff
@@ -64,9 +64,18 @@ local function make_config_info(config, bufnr)
...
+    local root_dir
+    local status = coroutine.resume(coroutine.create(function()
+      local status, err = pcall(function()
+        root_dir = config.get_root_dir(buffer_dir)
+      end)
+      if not status then
+        print('lspconfig: unhandled error: ' .. tostring(err))
+      end
+    end))
+    config_info.root_dir = status and root_dir or error_messages.root_dir_not_found
```
Comment (ii14) [inline lua/lspconfig/ui/lspinfo.lua:78]: If we don't support async root_dir functions right now then there should be at least an error message that says that
```suggestion
    local co = coroutine.create(function()
      local status, err = pcall(function()
        root_dir = config.get_root_dir(buffer_dir)
      end)
      if not status then
        print('lspconfig: unhandled error: ' .. tostring(err))
      end
    end)
    coroutine.resume(co)
    if root_dir then
      config_info.root_dir = root_dir
    elseif coroutine.status(co) == 'suspended' then
      config_info.root_dir = error_messages.async_root_dir_function
    else
      config_info.root_dir = error_messages.root_dir_not_found
    end
```
And add to `error_messages`:
```lua
  async_root_dir_function = 'Asynchronous root_dir functions are not supported in :LspInfo',
```

File: lua/lspconfig/server_configurations/rust_analyzer.lua [hunk 17]
```diff
@@ -10,6 +10,35 @@ local function reload_workspace(bufnr)
...
+  })
+
+  if jobid <= 0 then
+    vim.api.nvim_err_writeln('Failed to start cargo metadata job id:' .. jobid)
+    return
+  end
+
+  coroutine.yield()
+  local data = table.concat(chunks, '')
+  chunks = vim.json.decode(data)
```
Comment (ii14) [inline lua/lspconfig/server_configurations/rust_analyzer.lua:38]:
```suggestion
  chunks = vim.json.decode(table.concat(chunks, ''))
```

File: lua/lspconfig/configs.lua [hunk 18]
```diff
@@ -255,35 +248,47 @@ function configs.__newindex(t, config_name, config_def)
...
+        end
+        vim.schedule(function()
+          manager.add(root_dir, single_mode, bufnr)
+        end)
       end
+
+      util.async_run(function()
+        local root_dir
+        if get_root_dir then
+          root_dir = get_root_dir(buf_path, bufnr)
```
Comment (ii14) [inline lua/lspconfig/configs.lua:272]: Because `get_root_dir` can be asynchronous, buffer could be deleted in the meantime
```suggestion
          root_dir = get_root_dir(buf_path, bufnr)
          if not api.nvim_buf_is_valid(bufnr) then
            return
          end
```

File: lua/lspconfig/configs.lua [hunk 19]
```diff
@@ -97,54 +90,55 @@ function configs.__newindex(t, config_name, config_def)
...
+      if (#bufname == 0 and not config.single_file_support) or (#bufname ~= 0 and not util.bufname_valid(bufname)) then
+        return
+      end
+
+      local pwd = uv.cwd()
+
+      util.async_run(function()
+        local root_dir
+        if get_root_dir then
           root_dir = get_root_dir(util.path.sanitize(bufname), bufnr)
```
Comment (ii14) [inline lua/lspconfig/configs.lua:104]: Same here
```suggestion
          root_dir = get_root_dir(util.path.sanitize(bufname), bufnr)
          if not api.nvim_buf_is_valid(bufnr) then
            return
          end
```

File: lua/lspconfig/ui/lspinfo.lua (hunk 16 above)
Comment (ii14) [reply lua/lspconfig/ui/lspinfo.lua]: Actually, `root_dir` in rust looks so simple that it should be very easy to support both synchronous and asynchronous versions. If `coroutine.running()` is `nil` then you just call `vim.fn.jobwait`. Maybe for now `root_dir` functions should be required to support synchronous version just to keep `:LspInfo` working

File: lua/lspconfig/ui/lspinfo.lua [hunk 20]
```diff
@@ -64,9 +65,25 @@ local function make_config_info(config, bufnr)
...
-    config_info.root_dir = root_dir
+
+  if config.get_root_dir then
+    local root_dir
+    local co = coroutine.create(function()
+      local status, err = pcall(function()
+        root_dir = config.get_root_dir(buffer_dir)
+      end)
+      if not status then
+        print('lspconfig: unhandled error: ' .. tostring(err))
```
Comment (ii14) [inline lua/lspconfig/ui/lspinfo.lua:76]:
```suggestion
        vim.notify(('[lspconfig] unhandled error: %s'):format(tostring(err), vim.log.levels.WARN)
```

File: lua/lspconfig/util.lua [hunk 21]
```diff
@@ -588,4 +588,18 @@ function M.strip_archive_subpath(path)
...
 
+function M.async_run(fn)
+  vim.validate {
+    fn = { fn, 'f' },
+  }
+  local co = coroutine.create(function()
+    local status, err = pcall(fn)
+
+    if not status then
+      print('lspconfig: unhandled error: ' .. tostring(err))
```
Comment (ii14) [inline lua/lspconfig/util.lua:599]:
```suggestion
      vim.notify(('[lspconfig] unhandled error: %s'):format(tostring(err), vim.log.levels.WARN)
```

File: lua/lspconfig/server_configurations/rust_analyzer.lua [hunk 22]
```diff
@@ -10,6 +10,34 @@ local function reload_workspace(bufnr)
...
+  })
+
+  if jobid <= 0 then
+    vim.api.nvim_err_writeln('Failed to start cargo metadata job id:' .. jobid)
+    return
+  end
+
+  coroutine.yield()
+  chunks = vim.json.decode(table.concat(chunks, ''))
+  return chunks and chunks['workspace_root'] or nil
```
Comment (ii14) [inline lua/lspconfig/server_configurations/rust_analyzer.lua:38]:
```suggestion
  local stdout = {}
  local stderr = {}
  local jobid = vim.fn.jobstart(cmd, {
    on_stdout = function(_, data, _)
      stdout[#stdout + 1] = table.concat(data, '\n')
    end,
    on_stderr = function(_, data, _)
      stderr[#stderr + 1] = table.concat(data, '\n')
    end,
    on_exit = function()
      coroutine.resume(co)
    end,
    stdout_buffered = true,
    stderr_buffered = true,
  })

  if jobid <= 0 then
    vim.notify(('[lspconfig] cmd (%q) failed:\n%s'):format(table.concat(cmd, ' '), stderr), vim.log.levels.WARN)
    return
  end

  coroutine.yield()
  stdout = vim.json.decode(table.concat(stdout, ''))
  return stdout and stdout['workspace_root'] or nil
```

File: lua/lspconfig/configs.lua [hunk 23]
```diff
@@ -255,35 +276,38 @@ function configs.__newindex(t, config_name, config_def)
...
-      bufnr = bufnr or api.nvim_get_current_buf()
-      local buf_filetype = api.nvim_buf_get_option(bufnr, 'filetype')
-      if config.filetypes then
-        for _, filetype in ipairs(config.filetypes) do
-          if buf_filetype == filetype then
-            manager.try_add(bufnr)
+      async_run(function()
+        local root_dir
+        if get_root_dir then
+          root_dir = get_root_dir(buf_path, bufnr)
```
Comment (ii14) [inline lua/lspconfig/configs.lua:290]:
```suggestion
          root_dir = get_root_dir(buf_path, bufnr)
          reenter()
```

File: lua/lspconfig/configs.lua [hunk 24]
```diff
@@ -4,6 +4,30 @@ local tbl_deep_extend = vim.tbl_deep_extend
...
+    fn = { func, 'f' },
+  }
+  local co = coroutine.create(function()
+    local status, err = pcall(func)
+
+    if not status then
+      vim.notify(('[lspconfig] unhandled error: %s'):format(tostring(err)), vim.log.levels.WARN)
+    end
+  end)
+  coroutine.resume(co)
```
Comment (ii14) [inline lua/lspconfig/configs.lua:28]:
```suggestion
  coroutine.resume(coroutine.create(function()
    local status, err = pcall(func)
    if not status then
      vim.notify(('[lspconfig] unhandled error: %s'):format(tostring(err)), vim.log.levels.WARN)
    end
  end))
```

**Comment by d-e-s-o** [comment]
f01429048950195bcab6c430b430fc44b2d9e05a broke `rust-analyzer` usage for me.

**Comment by justinmk** [comment]
What problem does this PR solve? It is always helpful to state the problem being solved, in the PR description.

**Comment by glepnir** [comment]
before rust get `workspace_root` is sync . now support async and also make gopls get cache async . in future can support `root_pattern` async by using luv fs module. 

it works well for me for a while for rust anlyzer and other people can confirm it works.  please @d-e-s-o   create an issue if you with error output. before check your env first.

**Comment by d-e-s-o** [comment]
> please @d-e-s-o create an issue if you with error output. before check your env first.

What do you mean by "check your env first"? What should I be looking for, please? The problem is there is no output and I am not sure how to debug this. To the degree I can tell, all that changes is that `rust-analyzer` is never started (it's certainly not a running process as it was before the change).

**Comment by glepnir** [comment]
Issue template have a guide for how to use minconfig. And i can confrim it works well for many people?   No output mean not found? Again can't reproduce with a min config
